### ISSUE TITLE
clang static analyzer checker core.NullDereference: false positive and a wrong path note message 

### ISSUE THREAD

**Issue by Geoffrey1014** [llvm/llvm-project#58644]
I got a false positive error when compiling the following CMVE program:
```c
#include <stdio.h>

int *a(int i) {
  int *n = 0;
  return n;
}
int main() { 
  int d;
  int *e;
  int **f = &e;
  for (int g = 0; g < 3; g++)
    for (d = 2; d; d--) {
      printf("NPD_FLAG\n");
      *f = a(f == 0);
    }
}
```
Compiling the above code with Clang (trunk) with `-Xclang -analyzer-config -Xclang widen-loops=true  --analyze --analyzer-output text -Xclang -analyzer-display-progress ` in https://godbolt.org/z/96z6j5EMf result in :
```bash
ANALYZE (Syntax): <source> a : 12.4 ms
ANALYZE (Syntax): <source> main : 0.1 ms
ANALYZE (Path,  Inline_Regular): <source> main : 35.7 ms
<source>:14:10: warning: Dereference of null pointer (loaded from variable 'f') [core.NullDereference]
      *f = a(f == 0);
       ~ ^
<source>:11:3: note: Loop condition is true.  Entering loop body
  for (int g = 0; g < 3; g++)
  ^
<source>:12:5: note: Loop condition is true.  Entering loop body
    for (d = 2; d; d--) {
    ^
<source>:12:5: note: Loop condition is true.  Entering loop body
<source>:12:5: note: Loop condition is false. Execution continues on line 11
<source>:11:3: note: Loop condition is true.  Entering loop body
  for (int g = 0; g < 3; g++)
  ^
<source>:12:17: note: Value assigned to 'f'
    for (d = 2; d; d--) {
                ^
<source>:12:5: note: Loop condition is true.  Entering loop body
    for (d = 2; d; d--) {
    ^
<source>:14:14: note: Assuming 'f' is equal to null
      *f = a(f == 0);
             ^~~~~~
<source>:14:10: note: Dereference of null pointer (loaded from variable 'f')
      *f = a(f == 0);
       ~ ^
===-------------------------------------------------------------------------===
                                Analyzer timers
===-------------------------------------------------------------------------===
  Total Execution Time: 0.0053 seconds (0.0497 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
   0.0022 ( 62.9%)   0.0011 ( 62.9%)   0.0033 ( 62.9%)   0.0357 ( 71.8%)  Path exploration time
   0.0003 (  8.0%)   0.0001 (  8.0%)   0.0004 (  8.0%)   0.0125 ( 25.1%)  Syntax-based analysis time
   0.0010 ( 29.1%)   0.0005 ( 29.1%)   0.0015 ( 29.1%)   0.0015 (  3.1%)  Path-sensitive report post-processing time
   0.0035 (100.0%)   0.0018 (100.0%)   0.0053 (100.0%)   0.0497 (100.0%)  Total

1 warning generated.
Compiler returned: 0
```

First,  there exists proof in function `main` that `f` cannot be null but CSA takes the assumption " 'f' is equal to null " just because it sees  "f == 0" at line 14. 
Second,  the path notes contain a weird message :
```bash
<source>:12:17: note: Value assigned to 'f'
    for (d = 2; d; d--) {
```

There is no variable `f` in  the expression `for (d = 2; d; d--) {`

**Comment by steakhal** [comment]
Thanks for the report.

During widening CSA most likely saw a reference to the variable `f` part of the `f == 0` expression within the inner most loop.
Unfortunately it did not consider that the expression `f == 0` could never overwrite the value of `f`, so it assumed it gets overwritten by a fresh symbol.

Later, when it reaches the `f == 0` expression, it no longer knows that `f` refers to `&x`, hence it should be non-null; so it splits the path and we end up on a path where `f` is assumed to be null.
https://godbolt.org/z/qxGx3fvqa

```c++
void clang_analyzer_dump(int *p);
void clang_analyzer_printState();

int top(int x) {
  int result = 0;
  int *p = &x;
  clang_analyzer_dump(p); // &x
  for (int i = 0; i < 2; i++) {
    for (int j = 0; j < 1; j++) {
      clang_analyzer_dump(p); // SymReg{conj}
      p == 0; // bifurcate
      result = *p;
    }
  }
  return result;
}
```

This is definitely a bug in widening.
FYI, almost every option which is not enabled by default could have similar nuances. Turn them on by weighing in this factor.
That being said, I'm not sure how this will get prioritized, but IMO it would be awesome to see improvements in the loop-widening area.

**Comment by Geoffrey1014** [comment]
> Thanks for the report.
> 
> During widening CSA most likely saw a reference to the variable `f` part of the `f == 0` expression within the inner most loop. Unfortunately it did not consider that the expression `f == 0` could never overwrite the value of `f`, so it assumed it gets overwritten by a fresh symbol.
> 
> Later, when it reaches the `f == 0` expression, it no longer knows that `f` refers to `&x`, hence it should be non-null; so it splits the path and we end up on a path where `f` is assumed to be null. https://godbolt.org/z/qxGx3fvqa
> 
> ```c++
> void clang_analyzer_dump(int *p);
> void clang_analyzer_printState();
> 
> int top(int x) {
>   int result = 0;
>   int *p = &x;
>   clang_analyzer_dump(p); // &x
>   for (int i = 0; i < 2; i++) {
>     for (int j = 0; j < 1; j++) {
>       clang_analyzer_dump(p); // SymReg{conj}
>       p == 0; // bifurcate
>       result = *p;
>     }
>   }
>   return result;
> }
> ```
> 
> This is definitely a bug in widening. FYI, almost every option which is not enabled by default could have similar nuances. Turn them on by weighing in this factor. That being said, I'm not sure how this will get prioritized, but IMO it would be awesome to see improvements in the loop-widening area.

Thanks for your detailed reply.

I want to ask a few more questions.  
How many assumptions does CSA takes? More specifically, what assumptions does NPD checker take?  Does CSA assume every function as the top level function for analysis ? @martong told me enabling loop widening can make CSA analyze from `main`.  Can I specify the only analying entrance (e.g. `main`) for CSA? 

Is it okay for me to think that when there is a part of a function beyond what CSA can analyze (such as the limit on the  loop unrolling), then CSA makes some assumptions?  how many similar situations are there? Is there a document for a detail description for various situations?

**Comment by Geoffrey1014** [comment]
[#6 [SA_Bugs](https://github.com/Geoffrey1014/SA_Bugs)](https://github.com/Geoffrey1014/SA_Bugs/issues/6)

**Comment by Geoffrey1014** [comment]
reference
Geoffrey1014/SA_Bugs/#6
